### PR TITLE
Can end task instance from trigger

### LINKED ISSUE #31718: Add capability to exit tasks from trigger
### Description

We should be able to exit the "task" from the trigger.

We can already log from triggers so that's one hurdle out of the way.

The remaining hurdle I think is xcom, given that triggers run in asyncio.  

One solution that comes to mind is add a queue listener running in a thread that processes the “messages” and handles submitting to xcom.

Another would be use sqlalchemy's asyncio support. 

Sqlalchemy asyncio support would not help us though with custom xcom backend -- so maybe we either do just listener or we add an "optimized" path when not using custom xcom backend.

One implementation question is... would you just make it so that the trigger must call T`rigger.xcom_push` manually (which sends to this queue listener) or would you also want to make it possible, perhaps in an analogy with execute , possible to send the xcom payload(s) in the trigger event, or perhaps through a dedicated `Trigger.success` method?

I think that I’d probably prefer forcing Trigger writers to do the xcom push manually because … an interface to handle both the default return_value as well as handling multiple key-values isn’t quickly coming to mind.

cc: @jedcunningham @phanikumv @vandonr-amz 

### Use case/motivation

This would save users money or increase cluster bandwidth because for sensor-like deferrable tasks, no task needs to get spun up just to exit.

### Related issues

_No response_

### Are you willing to submit a PR?

- [X] Yes I am willing to submit a PR!

### Code of Conduct

- [X] I agree to follow this project's [Code of Conduct](https://github.com/apache/airflow/blob/main/CODE_OF_CONDUCT.md)

### PR DESCRIPTION
**Pull request by dstandish**
With this change, if desired, we can avoid having to resume the task on a worker.  We can end the task (in states success, failed, or skipped) from the trigger itself.

Closes: #31718

cc @Lee-W

### CODE REVIEW

File: airflow/models/trigger.py [hunk 1]
```diff
@@ -150,14 +154,31 @@ def submit_event(cls, trigger_id, event, session: Session = NEW_SESSION) -> None
...
+                if event.xcom_return:
+                    task_instance.xcom_push(key=XCOM_RETURN_KEY, value=event.xcom_return)
+                if event.other_xcom:
+                    for key, value in event.other_xcom.items():
+                        if key == XCOM_RETURN_KEY:
+                            log.warning(
+                                "Trigger yielded `other_xcom` with reserved key %s; ignoring. ti=%s",
+                                XCOM_RETURN_KEY,
+                                task_instance,
+                            )
```
Comment (jedcunningham) [inline airflow/models/trigger.py]:
```suggestion
                            )
                            continue
```

Missing this?

File: airflow/triggers/temporal.py [hunk 2]
```diff
@@ -68,9 +68,8 @@ async def run(self):
         while self.moment > timezone.utcnow():
             self.log.info("sleeping 1 second...")
             await asyncio.sleep(1)
-        # Send our single event and then we're done
-        self.log.info("yielding event with payload %r", self.moment)
-        yield TriggerEvent(self.moment)
+        self.log.info("Sensor time condition reached; marking task successful and exiting")
+        yield TaskSuccessEvent()
```
Comment (dstandish) [inline airflow/triggers/temporal.py]: here we have implemented for timedelta sensor

File: airflow/sensors/time_delta.py [hunk 3]
```diff
@@ -64,7 +64,7 @@ class TimeDeltaSensorAsync(TimeDeltaSensor):
     def execute(self, context: Context):
         target_dttm = context["data_interval_end"]
         target_dttm += self.delta
-        self.defer(trigger=DateTimeTrigger(moment=target_dttm), method_name="execute_complete")
+        self.defer(trigger=DateTimeTrigger(moment=target_dttm))
 
     def execute_complete(self, context, event=None):
```
Comment (jedcunningham) [inline airflow/sensors/time_delta.py]: Can't we toss this too then?

File: airflow/triggers/base.py [hunk 4]
```diff
@@ -119,3 +119,34 @@ def __eq__(self, other):
...
             return other.payload == self.payload
         return False
+
+
+class BaseTaskEndEvent(TriggerEvent):
+    """Base event class to end the task without resuming on worker."""
+
+    task_instance_state: TaskInstanceState
+
+    def __init__(self, *, xcom_return: Any = None, other_xcom: dict[str, Any] | None = None, **kwargs):
```
Comment (dstandish) [inline airflow/triggers/base.py]: this particular bit of the interface... open to suggestions on.... basically anything in this file there's a lot of ways we could've gone... but in particular the xcom_return vs other_xcom split and naming.

i thought having them explicit like this is better than having something "implicitly" go to `return_value` etc

File: airflow/sensors/time_delta.py (hunk 3 above)
Comment (dstandish) [reply airflow/sensors/time_delta.py]: yes we can i was thinking backcompat though? i dunno

File: airflow/models/trigger.py (hunk 1 above)
Comment (dstandish) [reply airflow/models/trigger.py]: if / else

i know people round here all about the early return / continue etc... but i think sometimes, and in particular here, readability can be aided with an else 🤷

Comment (jedcunningham) [reply airflow/models/trigger.py]: (There isn't an else though?)

Comment (dstandish) [reply airflow/models/trigger.py]: oh haha i didn't look closely enough. i was thinking you were talking about the outer `if`, you are right

File: airflow/exceptions.py [hunk 5]
```diff
@@ -347,7 +347,7 @@ def __init__(
         self,
         *,
         trigger,
-        method_name: str,
+        method_name: str = None,
```
Comment (uranusjr) [inline airflow/exceptions.py]: Why do we need to add this default? (And the one in `defer`)

Comment (dstandish) [reply airflow/exceptions.py]: I suppose we don't... but i was thinking that None would mean "don't come back".  What are you thinking, just supply None at the call site? Or sentinel?  A default of "don't come back" seems like a sensible default.

File: airflow/exceptions.py [hunk 6]
```diff
@@ -340,14 +340,14 @@ class TaskDeferred(BaseException):
     Signal an operator moving to deferred state.
 
     Special exception raised to signal that the operator it was raised from
-    wishes to defer until a trigger fires.
+    wishes to defer until a trigger fires
```
Comment (dstandish) [inline airflow/exceptions.py]:
```suggestion
    wishes to defer until a trigger fires.
```

File: airflow/models/taskinstance.py [hunk 7]
```diff
@@ -1693,6 +1693,8 @@ def _execute_task(self, context, task_orig):
             execute_callable = getattr(task_to_execute, self.next_method)
             if self.next_kwargs:
                 execute_callable = partial(execute_callable, **self.next_kwargs)
+        elif self.next_kwargs is not None:
+            raise AirflowException("Task is coming out of deferral without next_method specified.")
```
Comment (uranusjr) [inline airflow/models/taskinstance.py]: This should use a exception subclass instead. (Although I kind of feel we could just ignore `next_kwargs` when `next_method` is None and document that.)

Comment (dstandish) [reply airflow/models/taskinstance.py]: the reason I added this is because i was checking out the "debuggability" when user misconfigures this.  i.e. when user uses a trigger that does _not_ exit and user provides no method.

without adding this catch, the task will just run the execute method (which perhaps surprisingly results in an infinite loop of execute and deferral and back again).

but i think i will change this to make the intention clearer

File: airflow/triggers/temporal.py (hunk 2 above)
Comment (dstandish) [reply airflow/triggers/temporal.py]: ok so... there's a potential backcompat problem here

i think it's pretty unlikely but you never know

case is, user subclasses this operator and does some actual work in execute complete...  

do you think we need to worry about that @potiuk @uranusjr

Comment (dstandish) [reply airflow/triggers/temporal.py]: to me i think that is venturing into xkcd 1172 territory

Comment (dstandish) [reply airflow/triggers/temporal.py]: actually yeah... i think we discussed this a while back and agreed (though may not have gotten around to documenting)

our backcompat guarantee for operators is about their behavior, not the exact structure of all of their methods. we don't publish them as "frameworks" to build upon -- they are end products.

so i think the operators are not a concern, but potentially the trigger class itself could be a concern since it now exits the task instead of sending back to task.  so at most, if we need to do anything, we would need either to create another version of DateTimeTrigger which exits, or make it parameterizable.  but we would be free to update the opreators to use this new behavior since it does not affect overall behavior, even if it leaves a legacy method unused (execute_complete, that is).

Comment (dstandish) [reply airflow/triggers/temporal.py]: ok so ... i updated to make DateTimeTrigger be backward compatible. you use `exit_task=True` if you want it to exit instead of send back to task.

**Comment by dstandish** [comment]
ok i think this is ready for another look

File: airflow/models/trigger.py [hunk 8]
```diff
@@ -150,14 +154,32 @@ def submit_event(cls, trigger_id, event, session: Session = NEW_SESSION) -> None
...
         ):
-            # Add the event's payload into the kwargs for the task
-            next_kwargs = task_instance.next_kwargs or {}
-            next_kwargs["event"] = event.payload
-            task_instance.next_kwargs = next_kwargs
-            # Remove ourselves as its trigger
-            task_instance.trigger_id = None
-            # Finally, mark it as scheduled so it gets re-queued
-            task_instance.state = TaskInstanceState.SCHEDULED
+            if isinstance(event, BaseTaskEndEvent):
```
Comment (uranusjr) [inline airflow/models/trigger.py]: Can this logic live in the event class instead to eliminate the if-else check? TriggerEvent can have a `handle_submit` (or something) that implements the else block here, and `BaseTaskEndEvent` overrides.

File: airflow/models/trigger.py [hunk 9]
```diff
@@ -150,14 +154,32 @@ def submit_event(cls, trigger_id, event, session: Session = NEW_SESSION) -> None
...
+                # task will be marked with terminal state and will not resume on worker
+                task_instance.trigger_id = None
+                task_instance.state = event.task_instance_state
+                if event.xcom_return:
+                    task_instance.xcom_push(key=XCOM_RETURN_KEY, value=event.xcom_return)
+                if event.other_xcom:
```
Comment (uranusjr) [inline airflow/models/trigger.py]: Do we need to babysit the user here? I feel we can just have one single `event.xcoms` and allow the user to use `XCOM_RETURN_KEY` if they really want to. This key isn’t _that_ special anyway and we don’t block users from pushing to it manually in other places.

File: airflow/sensors/date_time.py [hunk 10]
```diff
@@ -86,11 +86,4 @@ class DateTimeSensorAsync(DateTimeSensor):
...
     def execute(self, context: Context):
-        self.defer(
-            trigger=DateTimeTrigger(moment=timezone.parse(self.target_time)),
-            method_name="execute_complete",
-        )
-
-    def execute_complete(self, context, event=None):
-        """Callback for when the trigger fires - returns immediately."""
-        return None
+        self.defer(trigger=DateTimeTrigger(moment=timezone.parse(self.target_time), exit_task=True))
```
Comment (uranusjr) [inline airflow/sensors/date_time.py]: Reading this I start to wonder if `end` is a better term than `exit`.

File: airflow/triggers/base.py [hunk 11]
```diff
@@ -109,7 +109,7 @@ class TriggerEvent:
     events.
     """
 
-    def __init__(self, payload: Any):
+    def __init__(self, payload: Any = None):
```
Comment (uranusjr) [inline airflow/triggers/base.py]: When is a null payload useful? I feel this can stay a required argument.

File: airflow/triggers/base.py [hunk 12]
```diff
@@ -119,3 +119,34 @@ def __eq__(self, other):
...
+
+
+class BaseTaskEndEvent(TriggerEvent):
+    """Base event class to end the task without resuming on worker."""
+
+    task_instance_state: TaskInstanceState
+
+    def __init__(self, *, xcom_return: Any | None = None, other_xcom: dict[str, Any] | None = None, **kwargs):
+        if "payload" in kwargs:
+            raise ValueError("Param 'payload' not supported for this class.")
```
Comment (uranusjr) [inline airflow/triggers/base.py]: Why is this needed? I imagine this should work:

```suggestion
    def __init__(self, *, xcom_return: Any | None = None, other_xcom: dict[str, Any] | None = None) -> None:
```

File: airflow/triggers/temporal.py [hunk 13]
```diff
@@ -84,5 +95,5 @@ class TimeDeltaTrigger(DateTimeTrigger):
     DateTimeTrigger class, since they're operationally the same.
     """
 
-    def __init__(self, delta: datetime.timedelta):
-        super().__init__(moment=timezone.utcnow() + delta)
+    def __init__(self, delta: datetime.timedelta, **kwargs):
+        super().__init__(moment=timezone.utcnow() + delta, **kwargs)
```
Comment (uranusjr) [inline airflow/triggers/temporal.py]: Since the only valid kwarg is `exit_task` let’s spell it out directly; it’s friendlier for editors.

File: airflow/triggers/temporal.py [hunk 14]
```diff
@@ -32,9 +32,13 @@ class DateTimeTrigger(BaseTrigger):
...
 
     The provided datetime MUST be in UTC.
+
+    :param moment: when to yield event
+    :param exit_task: whether the trigger should mark the task successful after time condition
+        reached or resume the task
     """
 
-    def __init__(self, moment: datetime.datetime):
+    def __init__(self, moment: datetime.datetime, exit_task=False):
```
Comment (uranusjr) [inline airflow/triggers/temporal.py]: Can be a good idea to make the new argument kwarg-only

File: airflow/models/trigger.py (hunk 8 above)
Comment (dstandish) [reply airflow/models/trigger.py]: sure

File: airflow/models/trigger.py (hunk 9 above)
Comment (dstandish) [reply airflow/models/trigger.py]: My concern is that discovering the specialness of `XCOM_RETURN_KEY` is hard.

In classic operators, it's just the return value.  same with taskflow.  It's the default when you pull xcom.  I would bet it's something that users are pretty likely to never know.

But splitting it out is a bit clunky.  So i'll go ahead and combine

File: airflow/sensors/date_time.py (hunk 10 above)
Comment (dstandish) [reply airflow/sensors/date_time.py]: sure

File: airflow/triggers/base.py (hunk 12 above)
Comment (dstandish) [reply airflow/triggers/base.py]: Because the payload is determined by the state / controlled by the class.  User should not submit a payload.

File: airflow/triggers/temporal.py (hunk 13 above)
Comment (dstandish) [reply airflow/triggers/temporal.py]: sure

File: airflow/triggers/temporal.py (hunk 14 above)
Comment (dstandish) [reply airflow/triggers/temporal.py]: sure

File: airflow/triggers/base.py (hunk 12 above)
Comment (dstandish) [reply airflow/triggers/base.py]: if we do not check, user might pass `payload` and it would be silently ignored

File: airflow/triggers/base.py (hunk 11 above)
Comment (dstandish) [reply airflow/triggers/base.py]: sure

**Comment by dstandish** [comment]
ok made all changes except checking for user-supplied 'payload'

File: airflow/triggers/base.py (hunk 12 above)
Comment (uranusjr) [reply airflow/triggers/base.py]: I don’t think so? The suggested change removed `payload` altogether and should raise TypeError when it is provided. Or is the issue you want to allow `**kwargs`? But I don’t see any subclass using that either.

Comment (dstandish) [reply airflow/triggers/base.py]: Ah i see what you're saying...  Well it's in the base class.  Would you not say it's probably a good idea to include kwargs in the subclasses for future compat?

**Comment by dstandish** [comment]
marking this draft for now because we have to deal with callbacks and email
